### PR TITLE
Input refactor

### PR DESCRIPTION
**Pull request by h1alexbel**
Refactored ```software.amazon.event.ruler.input``` package. Introducing new interfaces for existing classes to decouple them.

### CODE REVIEW

File: src/main/software/amazon/event/ruler/input/MatchTypeParser.java
Comment (baldawar) [inline src/main/software/amazon/event/ruler/input/MatchTypeParser.java:6]: please remove

File: src/main/software/amazon/event/ruler/input/DefaultParser.java
Comment (baldawar) [inline src/main/software/amazon/event/ruler/input/DefaultParser.java:16]: I don't think the renaming is necessary but if you are doing so, please check if there's any documentation (in code or Readme) that need to be updated as well. I've found IDEs to not be able to pick them all.

File: src/main/software/amazon/event/ruler/input/WildcardParser.java
Comment (baldawar) [inline src/main/software/amazon/event/ruler/input/WildcardParser.java:21]: utf8Bytes is more accurate here.

File: src/main/software/amazon/event/ruler/input/WildcardParser.java
Comment (baldawar) [inline src/main/software/amazon/event/ruler/input/WildcardParser.java:14]: curious if you've considered suppressing individual methods first? Ideally you want to suppress the smallest part of the code (and add a comment on why suppression is necessary).

File: src/main/software/amazon/event/ruler/input/EqualsIgnoreCaseParser.java
Comment (baldawar) [inline src/main/software/amazon/event/ruler/input/EqualsIgnoreCaseParser.java:23]: is `type` redundant here?

File: src/main/software/amazon/event/ruler/MatchType.java
Comment (baldawar) [inline src/main/software/amazon/event/ruler/MatchType.java:18]: Can you elaborate why we have `NO` here?

File: src/main/software/amazon/event/ruler/Constants.java
Comment (baldawar) [inline src/main/software/amazon/event/ruler/Constants.java:10]: I recommend whenever you throw an exception to also include a contextual message.

File: src/main/software/amazon/event/ruler/Task.java
Comment (baldawar) [inline src/main/software/amazon/event/ruler/Task.java:59]: is this duplicate of https://github.com/aws/event-ruler/pull/60/files

File: src/main/software/amazon/event/ruler/input/StringValueParser.java
Comment (baldawar) [inline src/main/software/amazon/event/ruler/input/StringValueParser.java:12]: Needs documentation (and supporting unit tests) to understand the context and expected behavior.

File: src/main/software/amazon/event/ruler/MatchType.java
Comment (h1alexbel) [reply src/main/software/amazon/event/ruler/MatchType.java:18]: Yeah, because in Parser's method #parse(MatchType type, String value) we need to pass MatchType, to prevent using [```null```](https://www.yegor256.com/2014/05/13/why-null-is-bad.html), I suggest to use [Null Object Pattern](https://www.baeldung.com/java-null-object-pattern) instead.

Comment (h1alexbel) [reply src/main/software/amazon/event/ruler/MatchType.java:18]: > Can you elaborate why we have `NO` here?

Yeah, because in Parser's method #parse(MatchType type, String value) we need to pass MatchType, to prevent using [```null```](https://www.yegor256.com/2014/05/13/why-null-is-bad.html), I suggest to use [Null Object Pattern](https://www.baeldung.com/java-null-object-pattern) instead.

File: src/main/software/amazon/event/ruler/input/WildcardParser.java
Comment (h1alexbel) [reply src/main/software/amazon/event/ruler/input/WildcardParser.java:14]: My bad, I'll remove that.

**Comment by h1alexbel** [comment]
Hey, @baldawar I've updated PR with the consistent pkg code style as you have. Please take a look.

File: src/main/software/amazon/event/ruler/MatchType.java
Comment (baldawar) [inline src/main/software/amazon/event/ruler/MatchType.java:18]: nit, formatting of comment is off.

File: src/main/software/amazon/event/ruler/RuleCompiler.java
Comment (baldawar) [inline src/main/software/amazon/event/ruler/RuleCompiler.java:36]: missed a message here.

File: src/main/software/amazon/event/ruler/input/ByteParser.java
Comment (baldawar) [inline src/main/software/amazon/event/ruler/input/ByteParser.java:3]: I'd recommend you also include a minor doc on what we expect to use this interface. nothing too fancy, but enough to offer direction for anyone that comes across this class.

File: src/main/software/amazon/event/ruler/input/StringValueParser.java
Comment (baldawar) [inline src/main/software/amazon/event/ruler/input/StringValueParser.java:3]: I think I need one more input on the inclusion of these parsers. @jonessha  do you have an opinions on either keeping / removing / changing interfaces?

**Comment by baldawar** [comment]
Thanks @h1alexbel . Mostly LGTM. I just want a second opinion on the interfaces from Shawn as it might affect something he's potentially working in recent future. If he has no objections by tomorrow, probably good to merge if you can address the minor call-outs I made.

File: src/main/software/amazon/event/ruler/input/StringValueParser.java
Comment (jonessha) [reply src/main/software/amazon/event/ruler/input/StringValueParser.java:3]: I'm not strongly opposed to this change, but at the same time, I don't really get the purpose of including all these interfaces. In my mind for how Ruler will evolve, these interfaces will serve no purpose. @h1alexbel - curious what benefit you see these interfaces providing as the code base evolves?

Comment (h1alexbel) [reply src/main/software/amazon/event/ruler/input/StringValueParser.java:3]: > I'm not strongly opposed to this change, but at the same time, I don't really get the purpose of including all these interfaces. In my mind for how Ruler will evolve, these interfaces will serve no purpose. @h1alexbel - curious what benefit you see these interfaces providing as the code base evolves?

Benefit is maintainability in the future and possible API extension via interfaces, and that all classes will be used by some special contract.

Comment (baldawar) [reply src/main/software/amazon/event/ruler/input/StringValueParser.java:3]: looks like this PR has stalled. since there's no blockers with any changes Shawn has staged in a different branch / repo; I don't mind merging this for now. We can always undo if we think this creates too much friction in future.

**Review by baldawar — APPROVED** [review]

**Comment by baldawar** [comment]
@h1alexbel  thanks for the PR. 

If you end up wanting to make more changes in future; if I can get an heads up, I'll make sure to keep some capacity on the side for your changes.

**Pull request merged**
